File: java/src/test/java/org/sammancoaching/conditionals/ExampleConditionalsTest.java
package org.sammancoaching.conditionals;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.assertEquals;

class ExampleConditionalsTest {

    @ParameterizedTest
    @CsvSource({
            "4, 3",
            "3, 5",
            "3, 1",
            "3, 4"
    })
    void invert_negative(int expected, int actual) {
        assertEquals(expected, ExampleConditionals.invert_negative(actual));
    }

    @ParameterizedTest
    @CsvSource({
            "4, 3",
            "3, 5",
            "3, 1",
            "3, 4"
    })
    void invert_positive(int expected, int actual) {
        assertEquals(expected, ExampleConditionals.invert_positive(actual));
    }

    @ParameterizedTest
    @CsvSource({
            "2, 1",
            "7, 10",
            "29, 30",
            "30, 0"
    })
    void redundant_else(int x, int expected) {
        assertEquals(expected, ExampleConditionals.redundant_else(x));
    }

    @ParameterizedTest
    @CsvSource({
            "2, 1",
            "7, 10",
            "29, 30",
            "30, 0"
    })
    void missing_else(int x, int expected) {
        assertEquals(expected, ExampleConditionals.missing_else(x));
    }

    @ParameterizedTest
    @ValueSource(ints = {5, 7})
    void deMorganAnd_true(int valid) {
        Assertions.assertTrue(ExampleConditionals.deMorganAnd(valid));
    }

    @ParameterizedTest
    @ValueSource(ints = {4, 8})
    void deMorganAnd_false(int invalid) {
        Assertions.assertFalse(ExampleConditionals.deMorganAnd(invalid));
    }

    @ParameterizedTest
    @ValueSource(ints = {5, 7})
    void deMorganOr_true(int valid) {
        Assertions.assertTrue(ExampleConditionals.deMorganOr(valid));
    }

    @ParameterizedTest
    @ValueSource(ints = {4, 8})
    void deMorganOr_false(int invalid) {
        Assertions.assertFalse(ExampleConditionals.deMorganOr(invalid));
    }

    @ParameterizedTest
    @CsvSource({
            "4, 3, 0",
            "7, 3, 0",
            "1, 3, 0",
            "3, 4, 7",
            "3, 11, 0"
    })
    void join_AND(int x, int y, int expected) {
        assertEquals(expected, ExampleConditionals.join_AND(x, y));
    }

    @ParameterizedTest
    @CsvSource({
            "4, 3, 0",
            "7, 3, 0",
            "1, 3, 0",
            "3, 4, 7",
            "3, 11, 0"
    })
    void split_AND(int x, int y, int expected) {
        assertEquals(expected, ExampleConditionals.split_AND(x, y));
    }

    @ParameterizedTest
    @CsvSource({
            "0, 3, true",
            "3, 6, true",
            "-1, 3, true",
            "-1, 5, false",
            "-1, 10, true"
    })
    void join_OR(int x, int y, boolean expected) {
        assertEquals(expected, ExampleConditionals.join_OR(x, y));
    }

    @ParameterizedTest
    @CsvSource({
            "0, 3, true",
            "3, 6, true",
            "-1, 3, true",
            "-1, 5, false",
            "-1, 10, true"
    })
    void split_OR(int x, int y, boolean expected) {
        assertEquals(expected, ExampleConditionals.split_OR(x, y));
    }

    @ParameterizedTest
    @CsvSource({
            "4, 3, 36",
            "7, 3, 63",
            "1, 3, 0",
            "3, 4, 0"
    })
    void split_statements(int x, int y, int expected) {
        assertEquals(expected, ExampleConditionals.split_statements(x, y));
    }

    @ParameterizedTest
    @CsvSource({
            "4, 3, 36",
            "7, 3, 63",
            "1, 3, 0",
            "3, 4, 0"
    })
    void join_statements(int x, int y, int expected) {
        assertEquals(expected, ExampleConditionals.join_statements(x, y));
    }

}

File: csharp/RefactorConditionals/RefactorConditionals.Tests/ExampleConditionalsTest.cs
using Xunit;

namespace RefactorConditionals.Tests
{
    public class ExampleConditionalsTest
    {
        [Theory]
        [InlineData(4, 3)]
        [InlineData(3, 5)]
        [InlineData(3, 1)]
        [InlineData(3, 4)]
        public void InvertNegative(int expected, int actual)
        {
            Assert.Equal(expected, ExampleConditionals.InvertNegative(actual));
        }

        [Theory]
        [InlineData(4, 3)]
        [InlineData(3, 5)]
        [InlineData(3, 1)]
        [InlineData(3, 4)]
        public void InvertPositive(int expected, int actual)
        {
            Assert.Equal(expected, ExampleConditionals.InvertPositive(actual));
        }

        [Theory]
        [InlineData(2, 1)]
        [InlineData(7, 10)]
        [InlineData(29, 30)]
        [InlineData(30, 0)]
        public void RedundantElse(int x, int expected)
        {
            Assert.Equal(expected, ExampleConditionals.RedundantElse(x));
        }

        [Theory]
        [InlineData(2, 1)]
        [InlineData(7, 10)]
        [InlineData(29, 30)]
        [InlineData(30, 0)]
        public void MissingElse(int x, int expected)
        {
            Assert.Equal(expected, ExampleConditionals.MissingElse(x));
        }

        [Theory]
        [InlineData(5)]
        [InlineData(7)]
        public void DeMorganAndTrue(int valid)
        {
            Assert.True(ExampleConditionals.DeMorganAnd(valid));
        }

        [Theory]
        [InlineData(4)]
        [InlineData(8)]
        public void DeMorganAndFalse(int invalid)
        {
            Assert.False(ExampleConditionals.DeMorganAnd(invalid));
        }

        [Theory]
        [InlineData(5)]
        [InlineData(7)]
        public void DeMorganOrTrue(int valid)
        {
            Assert.True(ExampleConditionals.DeMorganOr(valid));
        }

        [Theory]
        [InlineData(4)]
        [InlineData(8)]
        public void DeMorganOrFalse(int invalid)
        {
            Assert.False(ExampleConditionals.DeMorganOr(invalid));
        }

        [Theory]
        [InlineData(4, 3, 0)]
        [InlineData(7, 3, 0)]
        [InlineData(1, 3, 0)]
        [InlineData(3, 4, 7)]
        [InlineData(3, 11, 0)]
        public void JoinAnd(int x, int y, int expected)
        {
            Assert.Equal(expected, ExampleConditionals.JoinAnd(x, y));
        }

        [Theory]
        [InlineData(4, 3, 0)]
        [InlineData(7, 3, 0)]
        [InlineData(1, 3, 0)]
        [InlineData(3, 4, 7)]
        [InlineData(3, 11, 0)]
        public void SplitAnd(int x, int y, int expected)
        {
            Assert.Equal(expected, ExampleConditionals.SplitAnd(x, y));
        }

        [Theory]
        [InlineData(0, 3, true)]
        [InlineData(3, 6, true)]
        [InlineData(-1, 3, true)]
        [InlineData(-1, 5, false)]
        [InlineData(-1, 10, true)]
        public void JoinOr(int x, int y, bool expected)
        {
            Assert.Equal(expected, ExampleConditionals.JoinOr(x, y));
        }

        [Theory]
        [InlineData(0, 3, true)]
        [InlineData(3, 6, true)]
        [InlineData(-1, 3, true)]
        [InlineData(-1, 5, false)]
        [InlineData(-1, 10, true)]
        public void SplitOr(int x, int y, bool expected)
        {
            Assert.Equal(expected, ExampleConditionals.SplitOr(x, y));
        }

        [Theory]
        [InlineData(4, 3, 36)]
        [InlineData(7, 3, 63)]
        [InlineData(1, 3, 0)]
        [InlineData(3, 4, 0)]
        public void SplitStatements(int x, int y, int expected)
        {
            Assert.Equal(expected, ExampleConditionals.SplitStatements(x, y));
        }

        [Theory]
        [InlineData(4, 3, 36)]
        [InlineData(7, 3, 63)]
        [InlineData(1, 3, 0)]
        [InlineData(3, 4, 0)]
        public void JoinStatements(int x, int y, int expected)
        {
            Assert.Equal(expected, ExampleConditionals.JoinStatements(x, y));
        }
    }
}